src/header_map.rs

```diff
@@ -1,50 +1,173 @@
+    pub fn remove_i<T: Into<i128>>(&mut self, key: T) -> Option<Value> {
+        self.0.remove(&Key::Integer(key.into()))
+    }
+
+    /// Remove a claim value with a text key.
+    pub fn remove_t<T: Into<String>>(&mut self, key: T) -> Option<Value> {
+        self.0.remove(&Key::Text(key.into()))
     }
 }
 
-impl AsRef<BTreeMap<Value, Value>> for HeaderMap {
-    fn as_ref(&self) -> &BTreeMap<Value, Value> {
+impl AsRef<BTreeMap<Key, Value>> for HeaderMap {
+    fn as_ref(&self) -> &BTreeMap<Key, Value> {
         &self.0
     }
 }
 
 impl Deref for HeaderMap {
-    type Target = BTreeMap<Value, Value>;
+    type Target = BTreeMap<Key, Value>;
 
-    fn deref(&self) -> &BTreeMap<Value, Value> {
+    fn deref(&self) -> &BTreeMap<Key, Value> {
         &self.0
     }
 }
 
-impl From<BTreeMap<Value, Value>> for HeaderMap {
-    fn from(m: BTreeMap<Value, Value>) -> Self {
+impl FromIterator<(Key, Value)> for HeaderMap {
+    fn from_iter<T: IntoIterator<Item = (Key, Value)>>(iter: T) -> Self {
+        Self(BTreeMap::from_iter(iter))
+    }
+}
+
+impl TryFrom<BTreeMap<Value, Value>> for HeaderMap {
+    type Error = Error;
+
+    fn try_from(m: BTreeMap<Value, Value>) -> Result<Self, Self::Error> {
+        m.into_iter()
+            .map(|(k, v)| Ok((Key::try_from(k)?, v)))
+            .collect()
+    }
+}
+
+impl From<BTreeMap<Key, Value>> for HeaderMap {
+    fn from(m: BTreeMap<Key, Value>) -> Self {
         Self(m)
     }
 }
+
+impl IntoIterator for HeaderMap {
+    type Item = (Key, Value);
+
+    type IntoIter = <BTreeMap<Key, Value> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+pub trait Header: Into<Value> + TryFrom<Value, Error = serde_cbor::Error> {
+    fn key() -> Key;
+}
+
+/// Simple macro for defining generic claims with implementations of
+/// the Claim and From<> for serde_cbor::Value traits.
+/// Custom value_type's must implement From<value_type> for serde_cbor::Value.
+macro_rules! define_header {
+    ($name:ident, $value_type: ty, $key: expr) => {
+        #[derive(Clone, Debug, Deserialize, Serialize, PartialEq, PartialOrd, Eq, Ord, Hash)]
+        pub struct $name($value_type);
+        impl $name {
+            pub fn new(value: $value_type) -> $name {
+                $name(value)
+            }
+        }
+
+        impl Header for $name {
+            fn key() -> Key {
+                $key
+            }
+        }
+
+        impl From<$name> for Value {
+            fn from(value: $name) -> Self {
+                value.0.into()
+            }
+        }
+
+        impl TryFrom<Value> for $name {
+            type Error = serde_cbor::Error;
+
+            fn try_from(value: Value) -> Result<Self, Self::Error> {
+                Ok(Self(serde_cbor::value::from_value(value)?))
+            }
+        }
+    };
+}
+
+define_header!(Algorithm, String, Key::Integer(1));
+define_header!(X5Chain, Vec<u8>, Key::Integer(33));
```

Comment:
`X5Chain` is a bit awkward. If there is a single certificate in the chain it can be a ByteStr as you have defined, but if there is more than one in the chain it is an array of ByteStr's. So probably not doable in this macro.